### PR TITLE
`General`: Refine editor access rights

### PR DESCRIPTION
**Pull request by FlorianGlombik**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code.
- [x] I added multiple screenshots/screencasts of my UI changes.

### Motivation and Context

Based on the feedback during the summer semester 2021 the permissions of the editor role are slightly adjusted.

### Description

1. Editors are no longer able to delete _Lectures_, _Lecture Units_, and _Lecture Attachments_
2. Editors can now access the template and solution submissions for programming exercises
3. Editors can now access programming exercises in the exam mode _(currently they were able to access other exercises but not programming exercises - this should be consistent either one or the other way)_
4. Make sure that the behavior for instructors did not change

### Steps for Testing

Prerequisites:
- 1 Editor
- 1 Instructor

#### 1. Editors are no longer able to delete _Lectures_, _Lecture Units_, and _Lecture Attachments_
1. Navigate to _Lectures_
2. See that no delete buttons are displayed for lectures, lecture units and lecture attachments
3. Verify that the corresponding REST calls can not be used by editors anymore _(returning a 404)_

#### 2. Editors can now access the template and solution submissions for programming exercises
1. Navigate to the detail view of programming exercise
2. Make sure that editors and instructors can access the template and solution submissions
3. Make sure that editors can now access other submissions (student submissions)

#### 3. Editors can now access programming exercises in the exam mode
1. Navigate to the detail view of an exam programming exercise as an editor
2. Make sure that editors and instructors can access the template and solution submissions

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### 1. Editors are no longer able to delete _Lectures_, _Lecture Units_, and _Lecture Attachments_
<details>
  <summary>Lectures</summary>
Before:

image

After:
image

</details>
<details>
  <summary>Lecture Units</summary>
Before:

image

After:

image

</details>
<details>
  <summary>Lecture Attachments</summary>
Before:

image

After:

image

</details>

#### 2. Editors can now access the template and solution submissions for programming exercises
The page is now accessible for editors: 

image

#### 3. Editors can now access programming exercises in the exam mode
<details>
Before:

https://user-images.githubusercontent.com/63976129/140758632-78cde17f-5ed0-4c78-a62c-f871ae8f46c2.mp4

After:

https://user-images.githubusercontent.com/63976129/140767055-bba72fe7-daa6-4bcd-9854-9434d95e8b44.mp4

</details>

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 1]
```diff
@@ -137,7 +136,7 @@ public LectureResource(LectureRepository lectureRepository, LectureService lectu
         User user = userRepository.getUserWithGroupsAndAuthorities();
         Course course = courseRepository.findByIdElseThrow(courseId);
         if (!authCheckService.isAtLeastEditorInCourse(course, user)) {
-            return forbidden();
+            throw new AccessForbiddenException("You do not have sufficient permissions load lectures for course administration purposes!");
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:139]:
```suggestion
            throw new AccessForbiddenException("You do not have sufficient permissions to load lectures for course administration purposes!");
```

**Review by cremertim — APPROVED** [review]
I tested it on ts1 according to the testing steps

I can no longer see the delete button with an editor but can still see it with an instructor.

Editors are able to navigate into the submission page of the template commits aswell

The code looks good to me aswell

File: src/main/webapp/app/lecture/lecture.service.ts [hunk 2]
```diff
@@ -114,18 +113,33 @@ export class LectureService {
...
     }
 
+    /**
+     * Besides the within the lecture included variables for access rights the access rights of the
+     * respective course are set aswell.
+     *
+     * @param lecture for which the access rights shall be set
+     * @private
+     */
+    private setAccessRightsLecture(lecture: Lecture | null) {
```
Comment (Olimbia23) [inline src/main/webapp/app/lecture/lecture.service.ts:132]: [nit] you forgot the `@return` tag. But frankly, do we even need to return the `Lecture` in this method?
Can we make this method, be a method of class `Lecture`? As in:
```
class Lecture {
  ...
  setAccessRights() {
    ...
  }
}
```

**Review by Olimbia23 — APPROVED** [review]
Code changes look good.
Nice work @FlorianGlombik 
Approve 👍 

Ps. left one minor nit comment

File: src/main/java/de/tum/in/www1/artemis/web/rest/lecture/LectureUnitResource.java [hunk 3]
```diff
@@ -120,8 +122,8 @@ public LectureUnitResource(AuthorizationCheckService authorizationCheckService,
         if (!lectureUnit.getLecture().getId().equals(lectureId)) {
             return conflict();
         }
-        if (!authorizationCheckService.isAtLeastEditorInCourse(lectureUnit.getLecture().getCourse(), null)) {
-            return forbidden();
+        if (!authorizationCheckService.isAtLeastInstructorInCourse(lectureUnit.getLecture().getCourse(), null)) {
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/lecture/LectureUnitResource.java:125]: please use the variant `...elseThrow`. Then you don't need the if statement

Comment (FlorianGlombik) [reply src/main/java/de/tum/in/www1/artemis/web/rest/lecture/LectureUnitResource.java]: Thanks for the heads up, this is addressed in https://github.com/ls1intum/Artemis/pull/4201/commits/6f97a6ac86dd7cf84326cfaa594e8478061a2b4a 
I was not aware of the `checkHasAtLeastRoleInCourseElseThrow` method in the `AuthorizationCheckService`

File: src/main/webapp/app/lecture/lecture.service.ts (hunk 2 above)
Comment (FlorianGlombik) [reply src/main/webapp/app/lecture/lecture.service.ts]: Thanks for the review, the nitpick about the comment is fixed in https://github.com/ls1intum/Artemis/pull/4201/commits/5a069000921ad665e0170c134a70481c1d31fb53
Within this service the return value is not needed yet, it is done like that within other services so I decided to keep it consistent. 
I would not put this method into the class itself, as we are currently using those classes rather as models as far as I know. I think that the `AccountService` would be a place better place for such a method.

Comment (FlorianGlombik) [reply src/main/webapp/app/lecture/lecture.service.ts]: When refactoring this, the other services should be touched aswell _(services that have been touched in https://github.com/ls1intum/Artemis/pull/3742)_

**Review by b-fein — APPROVED** [review]
Code review: Looks good!

**Review by daniels98it — APPROVED** [review]
Tested on TS3 and looks good 👍

**Review by cremertim — APPROVED** [review]
re approve, tested on ts 1 again and it worked

**Review by MDK5 — APPROVED** [review]
Reviewed code in testing session, changes look good to me!

**Pull request merged**